### ISSUE TITLE
Locally build images can't be pulled or used as base image for another build

### ISSUE THREAD

**Issue by Lotti** [containers/podman#19856]
**TL;DR**: You don't need to pull a locally builded image. Remove `--pull` from the build command.

### Issue Description

Sorry for open a bug if is not, but I couldn't find any solution here or in documentation.

I'm running on Windows 11 with the podman default machine running in WSL2.

It seems that the local registry doesn't work outside the box without further configurations. It keeps to look at localhost:443. It doesn't work also with --tls-verify=false which reverts to port 80.

Following the isntructions reported [here](https://github.com/containers/podman/blob/main/troubleshooting.md#4-http-server-gave-http-response-to-https-client) to permanent add a registry doesn't work too (it seems that my file under `/etc/containers/registries.conf.d/` is ignored).

I read that using the tag "latest" to build a local image, makes podman always think as a remote repository, so I changed the tag from `latest` to `base`: nothing changed.

### Steps to reproduce the issue

Steps to reproduce the issue
1. install podman and podman desktop
2. `podman build -t localhost/baseimage:aaa .` with any Dockerfile
3. `podman pull --tls-verify=false localhost/baseimage:aaa`

### Describe the results you received

Here you'll see the end of the base image build and the start of the derivated build that always looks for localhost registry at port 443. I've enabled the trace logging to provide more details as possible.

```
COMMIT localhost/captcha-baseimage:base
--> 287d699b0fc9
Successfully tagged localhost/captcha-baseimage:base
287d699b0fc94af9f68e8d20304d3692b6b573159a29bab25431a1069eecef7f
PS> podman --log-level trace build --build-arg http_proxy=$http_proxy --build-arg https_proxy=$https_proxy --pull --rm --file Dockerfile -t localhost/captcha-machine:base .
time="2023-09-05T08:52:41+02:00" level=info msg="C:\\Program Files\\RedHat\\Podman\\podman.exe filtering at log level trace"
time="2023-09-05T08:52:41+02:00" level=debug msg="Called build.PersistentPreRunE(C:\\Program Files\\RedHat\\Podman\\podman.exe --log-level trace build --build-arg http_proxy= --build-arg https_proxy= --pull --rm --file Dockerfile -t localhost/captcha-machine:base .)"
time="2023-09-05T08:52:41+02:00" level=debug msg="SSH Ident Key \"C:\\\\Users\\\\Administrator\\\\.ssh\\\\podman-machine-default\" SHA256:WdXdB+SsDtCCNEtImge8CtpKNNsly+E0J4cIw0H7h4k ssh-ed25519"
time="2023-09-05T08:52:43+02:00" level=debug msg="DoRequest Method: GET URI: http://d/v4.6.2/libpod/_ping"
time="2023-09-05T08:52:43+02:00" level=debug msg="Loading registries configuration \"/etc/containers/registries.conf\""
time="2023-09-05T08:52:44+02:00" level=debug msg="DoRequest Method: POST URI: http://d/v4.6.2/libpod/build"
STEP 1/57: FROM localhost/captcha-baseimage:base
Trying to pull localhost/captcha-baseimage:base...
Error: creating build container: initializing source docker://localhost/captcha-baseimage:base: pinging container registry localhost: Get "https://localhost/v2/": dial tcp 127.0.0.1:443: connect: connection refused

time="2023-09-05T08:53:24+02:00" level=debug msg="Shutting down engines"
```

Same build but with `--tls-verify=false`
```
PS> podman --log-level trace build --tls-verify=false --build-arg http_proxy=$http_proxy --build-arg https_proxy=$https_proxy --pull --rm --file Dockerfile -t localhost/captcha-machine:base .
time="2023-09-05T09:15:17+02:00" level=info msg="C:\\Program Files\\RedHat\\Podman\\podman.exe filtering at log level trace"
time="2023-09-05T09:15:17+02:00" level=debug msg="Called build.PersistentPreRunE(C:\\Program Files\\RedHat\\Podman\\podman.exe --log-level trace build --tls-verify=false --build-arg http_proxy= --build-arg https_proxy= --pull --rm --file Dockerfile -t localhost/captcha-machine:base .)"
time="2023-09-05T09:15:17+02:00" level=debug msg="SSH Ident Key \"C:\\\\Users\\\\Administrator\\\\.ssh\\\\podman-machine-default\" SHA256:WdXdB+SsDtCCNEtImge8CtpKNNsly+E0J4cIw0H7h4k ssh-ed25519"
time="2023-09-05T09:15:18+02:00" level=debug msg="DoRequest Method: GET URI: http://d/v4.6.2/libpod/_ping"
time="2023-09-05T09:15:18+02:00" level=debug msg="Loading registries configuration \"/etc/containers/registries.conf\""
time="2023-09-05T09:15:18+02:00" level=debug msg="DoRequest Method: POST URI: http://d/v4.6.2/libpod/build"
STEP 1/57: FROM localhost/captcha-baseimage:base
Trying to pull localhost/captcha-baseimage:base...
Error: creating build container: initializing source docker://localhost/captcha-baseimage:base: pinging container registry localhost: Get "http://localhost/v2/": dial tcp 127.0.0.1:80: connect: connection refused

time="2023-09-05T09:16:14+02:00" level=debug msg="Shutting down engines"
```

### Describe the results you expected

I expect to be able to pull or reuse the locally build image

### podman info output

```yaml
podman version

Client:       Podman Engine
Version:      4.6.2
API Version:  4.6.2
Go Version:   go1.20.3
Git Commit:   5db42e86862ef42c59304c38aa583732fd80f178
Built:        Mon Aug 28 20:09:01 2023
OS/Arch:      windows/amd64

Server:       Podman Engine
Version:      4.6.1
API Version:  4.6.1
Go Version:   go1.19.10
Built:        Fri Aug 11 00:07:55 2023
OS/Arch:      linux/amd64
```

podman info
```
host:
  arch: amd64
  buildahVersion: 1.31.2
  cgroupControllers:
  - cpuset
  - cpu
  - cpuacct
  - blkio
  - memory
  - devices
  - freezer
  - net_cls
  - perf_event
  - net_prio
  - hugetlb
  - pids
  - rdma
  - misc
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: conmon-2.1.7-2.fc37.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.7, commit: '
  cpuUtilization:
    idlePercent: 91.2
    systemPercent: 3.05
    userPercent: 5.75
  cpus: 4
  databaseBackend: boltdb
  distribution:
    distribution: fedora
    variant: container
    version: "37"
  eventLogger: journald
  freeLocks: 2048
  hostname: vriva-p
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.15.90.1-microsoft-standard-WSL2
  linkmode: dynamic
  logDriver: journald
  memFree: 6210142208
  memTotal: 8177598464
  networkBackend: netavark
  networkBackendInfo:
    backend: netavark
    dns:
      package: aardvark-dns-1.7.0-1.fc37.x86_64
      path: /usr/libexec/podman/aardvark-dns
      version: aardvark-dns 1.7.0
    package: netavark-1.7.0-1.fc37.x86_64
    path: /usr/libexec/podman/netavark
    version: netavark 1.7.0
  ociRuntime:
    name: crun
    package: crun-1.8.6-1.fc37.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.8.6
      commit: 73f759f4a39769f60990e7d225f561b4f4f06bcf
      rundir: /run/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +LIBKRUN +WASM:wasmedge +YAJL
  os: linux
  pasta:
    executable: /usr/bin/pasta
    package: passt-0^20230625.g32660ce-1.fc37.x86_64
    version: |
      pasta 0^20230625.g32660ce-1.fc37.x86_64
      Copyright Red Hat
      GNU Affero GPL version 3 or later <https://www.gnu.org/licenses/agpl-3.0.html>
      This is free software: you are free to change and redistribute it.
      There is NO WARRANTY, to the extent permitted by law.
  remoteSocket:
    exists: true
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.0-8.fc37.x86_64
    version: |-
      slirp4netns version 1.2.0
      commit: 656041d45cfca7a4176f6b7eed9e4fe6c11e8383
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.3
  swapFree: 2147483648
  swapTotal: 2147483648
  uptime: 0h 3m 58.00s
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries:
  search:
  - docker.io
store:
  configFile: /usr/share/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphRootAllocated: 1081101176832
  graphRootUsed: 1840185344
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 5
  runRoot: /run/containers/storage
  transientStore: false
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 4.6.1
  Built: 1691705275
  BuiltTime: Fri Aug 11 00:07:55 2023
  GitCommit: ""
  GoVersion: go1.19.10
  Os: linux
  OsArch: linux/amd64
  Version: 4.6.1
```
```


### Podman in a container

No

### Privileged Or Rootless

None

### Upstream Latest Release

Yes

### Additional environment details

Running with Windows WSL 2

### Additional information

Additional information like issue happens only occasionally or issue happens with a particular architecture or on a particular setting

**Comment by vrothberg** [comment]
Does it work when you run the registry on port 5000?

**Comment by Lotti** [comment]
No luck :(

```
--> a85e3ab2a8ee
STEP 7/7: CMD ["bash"]
COMMIT localhost:5000/captcha-baseimage:base
--> 42013aeee918
Successfully tagged localhost:5000/captcha-baseimage:base
42013aeee918455c0b8bc730f25a95dae1ffb8cb782c8d02b5bf4b107c9818f7
PS C:\Users\Administrator\repos\Captcha_PHP> podman --log-level trace build --tls-verify=false --build-arg http_proxy=$http_proxy --build-arg https_proxy=$https_proxy --pull --rm --file Dockerfile -t localhost:5000/captcha-machine:base .
time="2023-09-05T09:38:15+02:00" level=info msg="C:\\Program Files\\RedHat\\Podman\\podman.exe filtering at log level trace"
time="2023-09-05T09:38:15+02:00" level=debug msg="Called build.PersistentPreRunE(C:\\Program Files\\RedHat\\Podman\\podman.exe --log-level trace build --tls-verify=false --build-arg http_proxy= --build-arg https_proxy= --pull --rm --file Dockerfile -t localhost:5000/captcha-machine:base .)"
time="2023-09-05T09:38:15+02:00" level=debug msg="SSH Ident Key \"C:\\\\Users\\\\Administrator\\\\.ssh\\\\podman-machine-default\" SHA256:LKwSRtw3xAlMWMG6n4HmPi3riqeAU009XkDEVDml3PA ssh-ed25519"
time="2023-09-05T09:38:16+02:00" level=debug msg="DoRequest Method: GET URI: http://d/v4.6.2/libpod/_ping"
time="2023-09-05T09:38:16+02:00" level=debug msg="Loading registries configuration \"/etc/containers/registries.conf\""
time="2023-09-05T09:38:16+02:00" level=debug msg="DoRequest Method: POST URI: http://d/v4.6.2/libpod/build"
PS C:\Users\Administrator\repos\Captcha_PHP> podman --log-level trace build --tls-verify=false --build-arg http_proxy=$http_proxy --build-arg https_proxy=$https_proxy --pull --rm --file Dockerfile -t localhost:5000/captcha-machine:base .
time="2023-09-05T09:39:00+02:00" level=info msg="C:\\Program Files\\RedHat\\Podman\\podman.exe filtering at log level trace"
time="2023-09-05T09:39:00+02:00" level=debug msg="Called build.PersistentPreRunE(C:\\Program Files\\RedHat\\Podman\\podman.exe --log-level trace build --tls-verify=false --build-arg http_proxy= --build-arg https_proxy= --pull --rm --file Dockerfile -t localhost:5000/captcha-machine:base .)"
time="2023-09-05T09:39:00+02:00" level=debug msg="SSH Ident Key \"C:\\\\Users\\\\Administrator\\\\.ssh\\\\podman-machine-default\" SHA256:LKwSRtw3xAlMWMG6n4HmPi3riqeAU009XkDEVDml3PA ssh-ed25519"
time="2023-09-05T09:39:01+02:00" level=debug msg="DoRequest Method: GET URI: http://d/v4.6.2/libpod/_ping"
time="2023-09-05T09:39:01+02:00" level=debug msg="Loading registries configuration \"/etc/containers/registries.conf\""
time="2023-09-05T09:39:01+02:00" level=debug msg="DoRequest Method: POST URI: http://d/v4.6.2/libpod/build"
STEP 1/57: FROM localhost:5000/captcha-baseimage:base
Trying to pull localhost:5000/captcha-baseimage:base...
Error: creating build container: initializing source docker://localhost:5000/captcha-baseimage:base: pinging container registry localhost:5000: Get "http://localhost:5000/v2/": dial tcp 127.0.0.1:5000: connect: connection refused

time="2023-09-05T09:39:34+02:00" level=debug msg="Shutting down engines"
```

**Comment by vrothberg** [comment]
Do you run a registry on that port?  `podman run -d -p 5000:5000 registry:2`?

**Comment by Lotti** [comment]
No, I don't explicitly run any registry on my machine. Should I?

Building an image locally, automatically creates an entry under "localhost" without any registry running.

PS> podman images
REPOSITORY                        TAG         IMAGE ID      CREATED         SIZE
localhost:5000/captcha-baseimage  base        42013aeee918  38 minutes ago  989 MB
quay.io/centos/centos             centos7     8652b9f0cb4c  2 years ago     212 MB

**Comment by vrothberg** [comment]
Maybe we should take a step back.

When building or tagging an image that is a short name (i.e., without a registry), Podman will prefix the image with `localhost`.  So `podman build -t foo:tag` creates an image `localhost/foo:tag`.

Docker always uses `docker.io` but we decided to not tie Podman to one registry only since there are many registries int he wild.

**Comment by Lotti** [comment]
I didn't use a short name to tag the image. Maybe I used a wrong parameter or I'm missing one.
```
podman --log-level trace build --tls-verify=false --build-arg http_proxy=$http_proxy --build-arg https_proxy=$https_proxy --pull --rm --file Dockerfile -t localhost/captcha-machine:base .
podman --log-level trace build --tls-verify=false --build-arg http_proxy=$http_proxy --build-arg https_proxy=$https_proxy --pull --rm --file Dockerfile -t localhost:5000/captcha-machine:base .
```

Anyway, I'd like to take more than one step back :D Can you point me to a documentation where is explained how to build and use images locally? If the only way is to run a local registry, I'll do.

Thanks for your time, your real-time answers are very appreciated.

**Comment by vrothberg** [comment]
> Can you point me to a documentation where is explained how to build and use images locally? 

Once built, you can use the image.  After `podman build` you should see the images in the local storage `podman images`.  No need to `podman pull` those images.

**Comment by Lotti** [comment]
So if images are reported correctly (at least you didn't notice anything wrong in the `podman images` command I've posted), why I can't use the from another Dockerfile that starts with `FROM localhost:5000/captcha-baseimage:base`?

Why the podman build command cannot reach the local registry?

**Comment by vrothberg** [comment]
If the local image is tagged `localhost/foo:tag` you can use that in another Dockerfile.  I just saw the `--pull` flag.  Leaving that out should let the build pass.

The issue then is that Podman will attempt to pull from `localhost/foo:tag` but there is no local registry; which would explain the errors you are seeing.

**Comment by Lotti** [comment]
It works! Thank you!

I thought that images where pulled only if not already present in local cache.

Final comment for future people:
Remove `--pull` from build command if you are going to use a local builded image.

**Comment by vrothberg** [comment]
Glad you got it working!  Thanks for reaching out.
